### PR TITLE
ci:Run CI tests in parallel

### PR DESCRIPTION
**Pull request by marcotc**
This PR leverages [CircleCI's parallelism option](https://circleci.com/docs/2.0/parallelism-faster-jobs/) to execute our `test` CI step in multiple containers. This reduces our total CI by around 10 minutes (from 35 to 25 minutes).

Thankfully there's no large change required in our CI file, as these parallel containers run inside a single job:
Screen Shot 2020-08-26 at 6 22 57 PM

### CODE REVIEW

**Comment by brettlangdon** [comment]
Is there a way for us to validate that every test is run on every CI run?

File: Rakefile [hunk 1]
```diff
@@ -184,562 +184,577 @@ task :'release:docs' => :docs do
...
 task :ci do
   if Gem::Version.new(RUBY_VERSION) < Gem::Version.new(Datadog::VERSION::MINIMUM_RUBY_VERSION)
     raise NotImplementedError, "Ruby versions < #{Datadog::VERSION::MINIMUM_RUBY_VERSION} are not supported!"
   elsif Gem::Version.new('2.0.0') <= Gem::Version.new(RUBY_VERSION) \
         && Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.1.0')
     # Main library
-    sh 'bundle exec rake test:main'
-    sh 'bundle exec rake spec:main'
-    sh 'bundle exec rake spec:contrib'
+    declare 'bundle exec rake test:main'
```
Comment (marcotc) [inline Rakefile:209]: I'm not 100 sold on this name change to `declare`.

I wanted to change it to name that displayed that we are not necessarily running the command inline, the moment it's defined, but that we've informed the testing framework that we'd like to test this command.

Whether it's executed inline, or in a different container is implementation detail.

**Comment by marcotc** [comment]
@brettlangdon now we do :) https://github.com/DataDog/dd-trace-rb/pull/1159

But besides coverage tracking, I can't think of a good mechanism to coordinate information across distinct containers that wouldn't feel artificial.
I think we should be able to work with our coverage tracking in order to prevent such regressions from happening.

File: Rakefile [hunk 2]
```diff
@@ -184,562 +184,577 @@ task :'release:docs' => :docs do
...
+# Declare a command for execution.
+# Jobs are parallelized if running in CI.
+def declare(*args)
+  total_executors = ENV.key?('CIRCLE_NODE_TOTAL') ? ENV['CIRCLE_NODE_TOTAL'].to_i : nil
+  current_executor = ENV.key?('CIRCLE_NODE_INDEX') ? ENV['CIRCLE_NODE_INDEX'].to_i : nil
+
+  if total_executors && current_executor && total_executors > 1
+    @execution_count ||= 0
+    @execution_count += 1
+    sh(*args) if @execution_count % total_executors == current_executor
```
Comment (ericmustin) [inline Rakefile:196]: What's the logic here exactly? I'm not sure I follow

**Review by ericmustin — APPROVED** [review]
Lgtm, left a small comment about a clarification for my own understanding, but i'm fine with `declare`

File: Rakefile (hunk 2 above)
Comment (marcotc) [reply Rakefile:196]: @ericmustin This method splits tests based on the total number of test executors.

If we have 30 Ruby tests total, 10 total executors (CI containers), and we are currently running on executor number 3, for example, `bundle exec rake ci` should only run tests where `test_num % 10 == 3)` which would translate to tests number 3, 13, and 23.

**Pull request merged**